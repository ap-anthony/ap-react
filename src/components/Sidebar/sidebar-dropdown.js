import React from 'react';
import styled from 'styled-components';


export default class SidebarDropdown extends React.Component {

    constructor(props) {
        super(props);
    }
    
    render() {
        const SidebarDropdownList = styled.ul`
            list-style-type: none;
            margin: 0;
            padding: 0 0 0 15px;
            display: ${this.props.expanded ? 'block' : 'none'};
        `;
        return (
            <SidebarDropdownList>{this.props.children}</SidebarDropdownList>
        );
    }

}