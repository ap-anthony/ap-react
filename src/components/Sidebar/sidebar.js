import React from 'react';
import styled from 'styled-components';
import { colors, assets, styles } from '../../theme';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import SidebarDropdown from './sidebar-dropdown';

const Icon = styled.span`
    width: 23px;
    display: inline-block;
    margin: 0 5px 0 3px;
`;

const EndIcon = styled.span`
    position: absolute;
    right: 0;
`;

const SidebarContainer = styled.div`
    width: 100%;
    height: 100%;
    overflow-y: auto;
`;

const SidebarLinksContainer = styled.div`
    padding: 0 15px;
`;

const SidebarLink = styled.button`
    width: 100%;
    text-align: left;
    height: 100%;
    border: none;
    outline: none;
    background-color: transparent;
    color: white;
    font-size: 1em;
    padding: 5px;
    cursor: pointer;
    font-size: 16px;
    position: relative;
    height: 34px;
`;

const SidebarHeader = styled.div`
    height: 65px;
    width: 100%;
    display: flex;
`;

const SidebarHeaderLogo = styled.div`
    width: 95px;
    height: 100%;
    background-image: url(${assets.ap_logo_white});
    background-position: center center;
    background-size: contain;
    background-repeat: no-repeat;
`;

const SidebarHeaderTitle = styled.div`
    flex: 1;
    display: flex;
    flex-direction: column;
    justify-content: center;
    font-size: 18pt;
    text-shadow: 4px 4px 4px rgba(0, 0, 0, 0.1);
    color: white;
`;

const SidebarDropdownItemLink = styled.button`
    width: 100%;
    color: ${colors.gray5};
    outline: none;
    border: 0;
    background-color: transparent;
    cursor: pointer;
    height: 34px;
    padding-left: 31px;
    text-align: left;
    border-left: 1px solid ${colors.teal2};
    position: relative;
    text-shadow: ${styles.text_shadow_subtle};

    &::before {
        content: ' ';
        width: 7px;
        height: 7px;
        position: absolute;
        left: -4px;
        top: 13px;
        background-color: ${colors.teal3};
        border-radius: 50%;
    }
`;

function SidebarDropdownItem(props) {
    return (
        <li>
            <SidebarDropdownItemLink data-tabid={props.tabId}>{props.children}</SidebarDropdownItemLink>
        </li>
    );
}

export default class Sidebar extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            collapsed: props.collapsed,
            menuLinksExpanded: {}
        };

        this.clickSidebarLink = this.clickSidebarLink.bind(this);
    }

    clickSidebarLink(clickedKey) {
        // Collapses all menus but expands the one that was clicked
        let menuLinksExpandedCopy = Object.keys(this.props.menuLinks).reduce((result, key) => {
            result[key] = key === clickedKey && !this.state.menuLinksExpanded[key];
            return result;
        }, {});
        this.setState({
            menuLinksExpanded: menuLinksExpandedCopy
        });
    }

    render() {
        return (
            <SidebarContainer>
                <SidebarHeader>
                    <SidebarHeaderLogo></SidebarHeaderLogo>
                    <SidebarHeaderTitle>AdmissionPros</SidebarHeaderTitle>
                </SidebarHeader>
                <SidebarLinksContainer>
                    {Object.keys(this.props.menuLinks).map((currentKey) => {
                        return (
                            <div key={currentKey}>
                                <SidebarLink onClick={() => this.clickSidebarLink(currentKey)}>
                                    <Icon>
                                        <FontAwesomeIcon icon={
                                            this.props.menuLinks[currentKey][0].icon ?  this.props.menuLinks[currentKey][0].icon.replace('fa-', '')
                                                                                    :  'info-circle'
                                        } />
                                    </Icon>
                                    {currentKey}
                                    <EndIcon>
                                        <FontAwesomeIcon icon="chevron-down" />
                                    </EndIcon>
                                </SidebarLink>
                                <SidebarDropdown expanded={this.state.menuLinksExpanded[currentKey]}>
                                    {this.props.menuLinks[currentKey].map((link => (
                                        <SidebarDropdownItem key={link.menu_id} tabId={link.id}>
                                            {link.caption}
                                        </SidebarDropdownItem>
                                    )))}
                                </SidebarDropdown>
                            </div>
                        );
                    })}
                </SidebarLinksContainer>
            </SidebarContainer>
        );  
    }
}