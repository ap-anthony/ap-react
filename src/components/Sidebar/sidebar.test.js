import React from 'react';
import Sidebar from './sidebar';
import { render, fireEvent, cleanup } from 'react-testing-library';
import renderer from 'react-test-renderer';

const menuLinks = {
    "Management Consoles": [],
    "Visit Management": [],
    "Campus Safety": [],
    "Document Management": [],
    "Applicant Missing Items": [],
    "Email Management": [],
    "Mailing Management": [],
    "Codes Management": [],
    "Report Tools": [],
    "Portal Management": [],
    "Data Imports & Exports": [],
    "System Management": [],
    "AdmissionPros Only": [],
};

test('Sidebar loads links when provided data.', () => {
    const { getByText } = render(
        <Sidebar collapsed={false} menuLinks={menuLinks} />
    );
    Object.keys(menuLinks).forEach(key => {
        expect(getByText(key)).toBeTruthy();
    })
});

test('Sidebar renders correctly.', () => {
    const tree = renderer
        .create(<Sidebar collapsed={false} menuLinks={menuLinks} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});