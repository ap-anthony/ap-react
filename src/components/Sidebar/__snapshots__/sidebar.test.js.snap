// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly. 1`] = `
<div
  className="sidebar"
>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Management Consoles
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Visit Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Campus Safety
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Document Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Applicant Missing Items
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Email Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Mailing Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Codes Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Report Tools
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Portal Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      Data Imports & Exports
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      System Management
    </span>
  </div>
  <div
    className="sidebar-link"
  >
    <span
      className="sidebar-link-text"
    >
      AdmissionPros Only
    </span>
  </div>
</div>
`;
